pkg/identityserver/gormstore/end_device_store.go

```diff
@@ -217,3 +217,38 @@ func (s *deviceStore) DeleteEndDevice(ctx context.Context, id *ttnpb.EndDeviceId
 	defer trace.StartRegion(ctx, "delete end device").End()
 	return s.deleteEntity(ctx, id)
 }
+
+var errUnknownField = errors.DefineInvalidArgument(
+	"unknown_field_update",
+	"unknown field update in end device batch update",
+)
+
+func batchUpdateLastSeenAtQuery(placeholders []string) string {
+	return fmt.Sprintf(`
+	WITH updates (application_id, device_id, last_seen_at) AS (VALUES %s)
+	UPDATE end_devices AS end_device
+	SET last_seen_at = updates.last_seen_at
+	FROM updates
+	WHERE end_device.application_id = updates.application_id
+	AND end_device.device_id = updates.device_id
```

Comment:
And if you still want that "only update to more recent timestamps", then I think you can add something like

```sql
AND (end_device.last_seen_at IS NULL OR end_device.last_seen_at < updates.last_seen_at)
```